### ISSUE TITLE
vulkan

### ISSUE THREAD

**Issue by YoSTEALTH** [pyglet/pyglet#267]
Is there any plans on supporting vulkan?

**Comment by einarf** [comment]
What kind of support are we talking about here?

**Comment by YoSTEALTH** [comment]
equivalent of "OpenGL" I suppose

**Comment by einarf** [comment]
Still don't understand what you are looking for here. I know what Vulkan is. What do you want pyglet do do?

**Comment by t-o-n-y-p** [comment]
I think he suggests rendering with Vulkan API if supported by a hardware. Great idea for pyglet 3.0 by the way :) right now we are focused on OpenGL 3 with pyglet 2.0

**Comment by YoSTEALTH** [comment]
Yes, what @t-o-n-y-p said.

Also if there is any plays to suppose async event managers? Say if you are loading an image right now it would block your event till its done loading or open a new thread? You could `img = await loadimage(path)`

I am not sure all these falls under pyglet, maybe it too much changes? maybe i am better off create my own package, not sure... figured i ask first since i like pyglet.

**Comment by HiImJulien** [comment]
>  maybe i am better off create my own project, not sure... 
@YoSTEALTH Was thinking the same lol.

This should be considered somewhat important. [Apple announced that it'll drop OpenGL in future versions of macOS](https://www.google.com/search?client=safari&rls=en&q=apple+opengl+deprecated&ie=UTF-8&oe=UTF-8). So, there will be a time, in which pyglet will not be useable on macOS anymore.

**Comment by einarf** [comment]
It does sounds like a pyglet 3.x thing like already stated.

I highly doubt the OpenGL api will disapper completely from MacOS any time soon, but it will surely be stuck on version 4.1 and won't be optimized or improved. Their drivers was never great in the first place 😄 

Let's at least hope it has support in the new ARM based macs. If this is not the case I think pyglet on macs are doomed for a long time.

Moving to vulkan means a complete rewrite of all the rendering. I also means all GPUs below DX11/GL4.1 are no longer supported. pyglet 2.x are just moving from 2.1 to 3.3.

**Comment by HiImJulien** [comment]
@einarf Another solution would be to use a graphics API wrapper like [sokol-gfx](https://github.com/floooh/sokol),
which wraps OpenGL (X11), DirectX (Win32) and Metal (macOS) under a thin layer. 

This would, however, require pyglet to either build some native binaries or have them shipped right with. The other solution would be, of course, to write a custom wrapper.

**Comment by ryannathans** [comment]
aren't new opengl versions implemented on top of vulkan?

**Comment by HiImJulien** [comment]
No, not necessarily. There might be implementations around, that use Vulkan under the hood. I think the best bet could also to expose an OpenGL ES interface and use MoltenGL (given the case, that pyglet receives a free license).

**Comment by YoSTEALTH** [comment]
Hmm... looks like OpenGL is being run on top of Vulkan using [Mesa](https://www.mesa3d.org/)

**Comment by HiImJulien** [comment]
@iperov SPIR-V is an open standard and there are several compilers targeting this bytecode format. One could, in theory, come up with a pure python compiler, that either compiles a DSL, a subset of python or even just (a subset) of python bytecode to SPIR-V.

This would be an extensive process, but it'd be doable.

**Comment by benmoran56** [comment]
Closing this due to age, but the upcoming pyglet v3.0 does take many steps in this direction. While there is no Vulkan support at launch, it is something that is becoming possible due to pyglet v3.0 having completely decoupled rendering backends. Version 3.0 will bring back support for some legacy backends (v2.1), WebGL, and now puts many pieces in place for eventual Vulkan support. 

Anyone interested in this in particular is encouraged to join the Discord (or open a new ticket) to discuss it.
